### ISSUE TITLE
asp-all-route-data must be IDictionary<string,object> or RouteValueDictionary

### ISSUE THREAD

**Issue by Remleo** [aspnet/Mvc#7087]
What if I need to pass values collection? For example:
```html
<a href="@Url.Action(...., new { Values = new [] { "Foo", "Bar" } })"
```

Is this possible with `asp-all-route-data`?
I think it can be `IDictionary<string,object>` or `RouteValueDictionary` instead of `IDictionary<string,string>`.

I can't use anonymous class because in my case keys must be with dot. Example: `FooFilter.Values`. 

Anonymous class declaration will produce compilation error:
```html
<a href="@Url.Action(...., new { FooFilter.Values = new [] { "Foo", "Bar" } })"
```
and `asp-all-route-data` can be used because it must be `IDictionary<string,object>`.

I found only one solution:
```html
<a href="@Url.Action(...)?FooFilter.Values=Foo&FooFilter.Values=Bar"
```
and this solution is horrible....

**Comment by Eilon** [comment]
Is there a reason you don't use `RouteValueDictionary` then?

**Comment by Remleo** [comment]
This Ok:
```html
<a href="@Url.Action("List", "FooController", new RouteValueDictionary { {"area", "FooArea"}, {"FooFilter.Values", new[]{"StatusA", "StatusB"}} })">
```

But it can't be used for `asp-all-route-data`. And changing type of `asp-all-route-data` to 'IDictionary<string,object>' couses problems with legacy code. So ASP-team will not agree to such changes I suppose...

**Comment by Remleo** [comment]
If `asp-all-route-data` have been  `IDictionary<string,object>` then code will be more nice:
```html
<a asp-action="FooAction" asp-route-status="@(new[]{ "StatusA", "StatusB" })"
```

**Comment by grahamehorner** [comment]
the asp-route-data is an IDictionary<string,string> because URIs have no concept of data types, so the question is if your not able to use anonymous data types (string:string) how to serialise/encode the values you have in such a way that they give you a IDictionary<string,string> and some metadata? about the types; which of cause leads to the question of deserialisation by the receiving party into the correct type?

**Comment by Remleo** [comment]
If you look at the implementaion of `AnchorTagHelper` then you will see that `_routeValues` is passing to `RouteValueDictionary` contructor. That's all. There is no special handling. And `RouteValueDictionary` is `IDictionary<string,object>`!

```cs
            RouteValueDictionary routeValues = null;
            if (_routeValues != null && _routeValues.Count > 0)
            {
                routeValues = new RouteValueDictionary(_routeValues);
            }
```

and then
```cs
tagBuilder = Generator.GenerateActionLink(.... routeValues: routeValues, ...);
```
From source:
https://github.com/aspnet/Mvc/blob/c27b07ef3f8b07dfe1d1de38a8d7bf2d6a9298f4/src/Microsoft.AspNetCore.Mvc.TagHelpers/AnchorTagHelper.cs

Why this is a problem for `AnchorTagHelper.RouteValues` but isn't for `RouteValueDictionary`? Problematic is one.

**Comment by grahamehorner** [comment]
@Remleo yes your correct; however the RouteValueDictionary not just for creating URI links; and the boxing/unboxing the is performed at a low level by the appropriate urlHelper; by using a IDictionary<string,string> for the AnchorTagHelper this prevents a non-string value from casing a problem deeper in the stack when generating the tags url

**Comment by Remleo** [comment]
@grahamehorner In my view, the tag helpers are replacement of IUrlHelper's extensions. In `Url.Action(...)` I can use code like this: `new { FooFilter = new [] { "Foo", "Bar" } }`, but in `asp-route-*` I can't. And it is a bit confusing. This synthetic constraint is illogical as for me.

In my project I'm added own TagHelper with `n-route-*` property with type `IDictionary<string,object>` and yet no problems. In what cases it can cause problems?

**Comment by mkArtakMSFT** [comment]
@NTaylorMullen, are you aware of any scenarios, where the tag helper introduced by @Remleo can cause issues?

**Comment by NTaylorMullen** [comment]
> In what cases it can cause problems?

If you're looking to provide a `string` value for your route value you will end up having issues. You'll need to save your string value to a variable before setting it to the attribute.

**Comment by MRT-77** [comment]
`@{
            IDictionary<string, string> routeData = new Dictionary<string, string>();
            routeData.Add("data1", "12345");
            routeData.Add("data2", "67890");
}
<li><a asp-controller="Home" asp-action="About" asp-all-route-data="routeData">Test Route Data</a></li>`
This method is using `asp-all-route-data`.

**Comment by Remleo** [comment]
@MRT-77 hav you read my first comment? Your example is too simple. What if I need to pass collection as value?

@NTaylorMullen Why do you see the problems  in `AnchorTagHelper.RouteValues` but not in the `RouteValueDictionary`. AnchorTagHelper just uses `RouteValueDictionary` inside.

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen Why do you see the problems in AnchorTagHelper.RouteValues but not in the RouteValueDictionary? AnchorTagHelper just uses RouteValueDictionary inside.

It's an artifact of how the TagHelper system tries to associate HTML with the AnchorTagHelper.RouteValues property. Going to deep dive a bit here but hopefully it helps understand the problem.

When you write:
```
@{
   IDictionary<string, string> routeData = new Dictionary<string, string>();
}
<a asp-controller="Home" asp-all-route-data="routeData">...</a>
<a asp-controller="Home" asp-route-foo="something">...</a>
```

Behind the scenes Razor generates something like:
```C#
IDictionary<string, string> routeData = new Dictionary<string, string>();
tagHelper = new AnchorTagHelper();
tagHelper.Controller = "Home";
tagHelper.RouteValues = routeData;
Write(tagHelper);

tagHelper = new AnchorTagHelper();
tagHelper.Controller = "Home";
tagHelper.RouteValues["foo"] = "something";
```

Razor does this because it has a special understanding of `Dictionary<string, TType>`. When a TagHelper has a `Dictionary<string, TType>` the system enables users to add single entries into that dictionary or set the entire thing (as shown by the example above). Now imagine if `RouteValues` was exposed as `Dictionary<string, object>`; in that world the TagHelper system allows users to provide anything other than strings. For example:

```
@{
    var something = true;
}
<my-taghelper-that-takes-string-object asp-foo-key="something" />
```

would result in:
```C#
tagHelper = new MyTaghelperThatTakesStringObject();
tagHelper.Foo["key"] = something;
```

Note the lacking quotations around `something`. Basically, since HTML uses quotations to surround their attribute values but those quotations are also significant in a C# landscape we chose to have the route data represented as the most common case: a user doing string->string route key->values.

As a total side note, it would technically be possible to provide a <string,object> attribute but that would require having two dictionaries and two sets of attributes on the TagHelper. Something we believe isn't a good user experience.

**Comment by dougbu** [comment]
This seems to be a duplicate of closed issue #7826
